File: amazon-scraper/python/sellerOffers.py
import urllib.parse
import csv
import re
import requests
from bs4 import BeautifulSoup

# Configuration
token = "<SDO-token>"
asin = "B0DZDBWM5B"  # Change this to any product ASIN
geocode = "us"

# Build seller offers URL from ASIN
target_url = f"https://www.amazon.com/gp/product/ajax/aodAjaxMain/?asin={asin}"

# Make API request
targetUrl = urllib.parse.quote(target_url, safe="")
apiUrl = "https://api.scrape.do/?token={}&url={}&geoCode={}".format(token, targetUrl, geocode)
response = requests.request("GET", apiUrl)

soup = BeautifulSoup(response.text, "html.parser")

offers = []

for sold_by in soup.find_all("div", id="aod-offer-soldBy"):
    # Find parent container with price
    container = sold_by.parent
    while container and not container.find("span", class_="a-price-whole"):
        container = container.parent
    if not container:
        continue

    # Price
    whole = container.find("span", class_="a-price-whole")
    frac = container.find("span", class_="a-price-fraction")
    price = f"${whole.text.strip('.')}.{frac.text.strip()}" if whole else "N/A"

    # Seller
    seller_link = sold_by.find("a", href=lambda x: x and "/gp/aag/main" in x)
    seller = seller_link.text.strip() if seller_link else "Amazon.com"

    # Seller rating
    rating_div = sold_by.find("div", id="aod-offer-seller-rating")
    rating_text = " ".join(rating_div.find("span", class_="a-icon-alt").get_text().split()) if rating_div and rating_div.find("span", class_="a-icon-alt") else ""
    count_elem = rating_div.find("span", id=lambda x: x and "seller-rating-count" in str(x)) if rating_div else None
    count_text = " ".join(count_elem.get_text().split()) if count_elem else ""

    # Extract rating value
    rating_match = re.search(r"(\d+\.?\d*) out of 5", rating_text)
    seller_rating = rating_match.group(1) if rating_match else "N/A"

    # Extract rating count
    count_match = re.search(r"\((\d[\d,]*)\s*ratings\)", count_text)
    rating_count = count_match.group(1) if count_match else "N/A"

    # Extract positive percentage
    positive_match = re.search(r"(\d+)%\s*positive", count_text)
    positive_pct = f"{positive_match.group(1)}%" if positive_match else "N/A"

    # Ships from
    ships_div = container.find("div", id="aod-offer-shipsFrom")
    ships_elem = ships_div.find("span", class_="a-color-base") if ships_div else None
    ships_from = " ".join(ships_elem.get_text().split()) if ships_elem else "N/A"

    # Condition
    condition_div = container.find("div", id="aod-offer-heading")
    condition = " ".join(condition_div.get_text().split()) if condition_div else "New"

    # Delivery
    delivery_div = container.find("div", id=lambda x: x and "DELIVERY" in str(x).upper())
    delivery = " ".join(delivery_div.get_text().split())[:80] if delivery_div else "N/A"

    offers.append({
        "ASIN": asin,
        "Price": price,
        "Condition": condition,
        "Seller": seller,
        "Seller Rating": seller_rating,
        "Rating Count": rating_count,
        "Positive": positive_pct,
        "Ships From": ships_from,
        "Delivery": delivery
    })

print(f"Found {len(offers)} seller offers for ASIN: {asin}")

# Export to CSV
csv_file = "sellerOffers.csv"
headers = ["ASIN", "Price", "Condition", "Seller", "Seller Rating", "Rating Count", "Positive", "Ships From", "Delivery"]

with open(csv_file, "w", newline="", encoding="utf-8") as f:
    writer = csv.DictWriter(f, fieldnames=headers)
    writer.writeheader()
    writer.writerows(offers)

print(f"Data exported to {csv_file}")

File: amazon-scraper/node.js/sellerOffers.js
const axios = require('axios');
const cheerio = require('cheerio');
const createCsvWriter = require('csv-writer').createObjectCsvWriter;

// Configuration
const token = "<SDO-token>";
const asin = "B0DZDBWM5B"; // Change this to any product ASIN
const geocode = "us";

// Build seller offers URL from ASIN
const targetUrl = `https://www.amazon.com/gp/product/ajax/aodAjaxMain/?asin=${asin}`;

// Make API request
const encodedUrl = encodeURIComponent(targetUrl);
const apiUrl = `https://api.scrape.do/?token=${token}&url=${encodedUrl}&geoCode=${geocode}`;

async function scrapeSellerOffers() {
    const response = await axios.get(apiUrl);
    const $ = cheerio.load(response.data);

    const offers = [];

    $('#aod-offer-soldBy').each((index, element) => {
        const soldBy = $(element);

        // Find parent container with price
        let container = soldBy.parent();
        while (container.length && !container.find('span.a-price-whole').length) {
            container = container.parent();
        }
        if (!container.length) return;

        // Price
        const whole = container.find('span.a-price-whole').first();
        const frac = container.find('span.a-price-fraction').first();
        const price = whole.length ? `$${whole.text().replace('.', '')}.${frac.text().trim()}` : 'N/A';

        // Seller
        const sellerLink = soldBy.find('a[href*="/gp/aag/main"]').first();
        const seller = sellerLink.length ? sellerLink.text().trim() : 'Amazon.com';

        // Seller rating
        const ratingDiv = soldBy.find('#aod-offer-seller-rating');
        const ratingAlt = ratingDiv.find('span.a-icon-alt').text().trim();
        const countElem = ratingDiv.find('span[id*="seller-rating-count"]');
        const countText = countElem.text().trim();

        // Extract rating value
        const ratingMatch = ratingAlt.match(/(\d+\.?\d*) out of 5/);
        const sellerRating = ratingMatch ? ratingMatch[1] : 'N/A';

        // Extract rating count
        const countMatch = countText.match(/\((\d[\d,]*)\s*ratings\)/);
        const ratingCount = countMatch ? countMatch[1] : 'N/A';

        // Extract positive percentage
        const positiveMatch = countText.match(/(\d+)%\s*positive/);
        const positivePct = positiveMatch ? `${positiveMatch[1]}%` : 'N/A';

        // Ships from
        const shipsDiv = container.find('#aod-offer-shipsFrom');
        const shipsElem = shipsDiv.find('span.a-color-base').first();
        const shipsFrom = shipsElem.length ? shipsElem.text().trim().replace(/\s+/g, ' ') : 'N/A';

        // Condition
        const conditionDiv = container.find('#aod-offer-heading');
        const condition = conditionDiv.length ? conditionDiv.text().trim().replace(/\s+/g, ' ') : 'New';

        // Delivery
        const deliveryDiv = container.find('div[id*="DELIVERY"]').first();
        let delivery = deliveryDiv.length ? deliveryDiv.text().trim().replace(/\s+/g, ' ') : 'N/A';
        delivery = delivery.substring(0, 80);

        offers.push({
            ASIN: asin,
            Price: price,
            Condition: condition,
            Seller: seller,
            'Seller Rating': sellerRating,
            'Rating Count': ratingCount,
            Positive: positivePct,
            'Ships From': shipsFrom,
            Delivery: delivery
        });
    });

    console.log(`Found ${offers.length} seller offers for ASIN: ${asin}`);

    // Export to CSV
    const csvWriter = createCsvWriter({
        path: 'sellerOffers.csv',
        header: [
            { id: 'ASIN', title: 'ASIN' },
            { id: 'Price', title: 'Price' },
            { id: 'Condition', title: 'Condition' },
            { id: 'Seller', title: 'Seller' },
            { id: 'Seller Rating', title: 'Seller Rating' },
            { id: 'Rating Count', title: 'Rating Count' },
            { id: 'Positive', title: 'Positive' },
            { id: 'Ships From', title: 'Ships From' },
            { id: 'Delivery', title: 'Delivery' }
        ]
    });

    await csvWriter.writeRecords(offers);
    console.log('Data exported to sellerOffers.csv');
}

scrapeSellerOffers().catch(console.error);
